### PR TITLE
3D Object Detection Evaluation

### PR DESCRIPTION
**Pull request by benjaminrwilson**
## PR Summary

- Adds 3D Object Detection Evaluation API

## Testing

In order to ensure this PR works as intended, it is:

* [x] unit tested.
* [ ] other or not applicable (*additional detail/rationale required*)

## Compliance with Standards

As the author, I certify that this PR conforms to the following standards:

* [x] Code changes conform to [PEP8](https://www.python.org/dev/peps/pep-0008) and docstrings conform to the Google Python [style guide](https://google.github.io/styleguide/pyguide.html?showone=Comments#38-comments-and-docstrings).
* [x] A well-written summary explains what was done and why it was done.
* [x] The PR is adequately tested and the testing details and links to external results are included.

### CODE REVIEW

File: tests/evaluation/detection/test_eval.py [hunk 1]
```diff
@@ -0,0 +1,349 @@
+# <Copyright 2022, Argo AI, LLC. Released under the MIT license.>
+
+"""Detection evaluation unit tests.
+
+Only the last two unit tests here use map ROI information.
+The rest apply no filtering to objects that have their corners located outside of the ROI).
```
Comment (jhonykaesemodel) [inline tests/evaluation/detection/test_eval.py:6]: Delete `)`

File: tests/evaluation/detection/test_eval.py [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+        assert np.all(dts[:, :4])
+
+        # Check that all error metrics are zero.
+        assert (dts[:, 4:7] == 0).all()
+
+        # # Check that there are 2 regular vehicles.
+        # assert gts["category"].value_counts()["REGULAR_VEHICLE"] == 2
+
+        # # Check that there are no other labels.
+        # assert gts["category"].value_counts().sum() == 2
```
Comment (jhonykaesemodel) [inline tests/evaluation/detection/test_eval.py:182]: Won't we use these?

File: tests/evaluation/detection/test_eval.py [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+#     dts = annotations.copy()
+#     dts["score"] = 1.0
+#     annotations.loc[:, "num_interior_pts"] = 1
+
+#     detection_cfg = DetectionCfg(eval_only_roi_instances=True, max_num_dts_per_category=1000, dataset_dir=root_dir)
+#     dts_, gts_, metrics_ = evaluate(dts, annotations, detection_cfg)
+
+
+# if __name__ == "__main__":
+#     test_val_identity()
```
Comment (jhonykaesemodel) [inline tests/evaluation/detection/test_eval.py:349]: Are we going to use this?

File: tests/evaluation/detection/test_eval.py [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+
+    timestamp_to_egoposes = read_city_SE3_ego(log_dir)
+    params = annotations.loc[:, ORDERED_CUBOID_COL_NAMES]
+
+    avm = ArgoverseStaticMap.from_map_dir(log_dir / "map", build_raster=True)
+
+    
+    avm.raster_roi_layer.array
+    breakpoint()
+    # read_city_SE3_ego()
```
Comment (jhonykaesemodel) [inline tests/evaluation/detection/test_eval.py:324]: remove breakpoint and the commented function?

File: src/av2/evaluation/detection/utils.py [hunk 5]
```diff
@@ -0,0 +1,439 @@
...
+    permutation: NDArrayInt = np.argsort(-scores).tolist()
+    dts = dts[permutation]
+
+    is_evaluated_dts: NDArrayBool = np.ones(N, dtype=bool)
+    is_evaluated_gts: NDArrayBool = np.ones(M, dtype=bool)
+    if avm is not None and city_SE3_ego is not None:
+        is_evaluated_dts &= compute_objects_in_roi_mask(dts, city_SE3_ego, avm)
+        is_evaluated_gts &= compute_objects_in_roi_mask(gts, city_SE3_ego, avm)
+
+    # num_interior_pts: NDArrayFloat = gts[..., -1]
```
Comment (jhonykaesemodel) [inline src/av2/evaluation/detection/utils.py:129]: do we need this?

File: src/av2/evaluation/detection/utils.py [hunk 6]
```diff
@@ -0,0 +1,439 @@
...
+def compute_evaluated_dts_mask(
+    dts: NDArrayFloat,
+    cfg: DetectionCfg,
+) -> NDArrayBool:
+    """Compute the evaluated cuboids mask.
+
+    Valid cuboids meet _two_ conditions:
+        1. The cuboid's centroid (x,y,z) must lie within the maximum range in the detection configuration.
+        2. The cuboid must have at _least_ one point in its interior.
+
```
Comment (jhonykaesemodel) [inline src/av2/evaluation/detection/utils.py:361]: Docstring might be a bit confusing as we are not using the number of points condition in this function. Or should we add `is_valid_num_points: NDArrayBool = gts[:, -1] > 0` here too?

File: src/av2/evaluation/detection/utils.py [hunk 7]
```diff
@@ -0,0 +1,439 @@
...
+    affinity_thresholds_m: Tuple[float, ...] = (0.5, 1.0, 2.0, 4.0)
+    affinity_type: AffinityType = AffinityType.CENTER
+    categories: Tuple[str, ...] = tuple(x.value for x in CompetitionCategories)
+    dataset_dir: Optional[Path] = None
+    eval_only_roi_instances: bool = True
+    filter_metric: FilterMetricType = FilterMetricType.EUCLIDEAN
+    max_num_dts_per_category: int = 100
+    max_range_m: float = 200.0
+    num_recall_samples: int = 101
```
Comment (jhonykaesemodel) [inline src/av2/evaluation/detection/utils.py:71]: I think it's cleaner if we use 100 instead and then use `np.linspace(0, 1, cfg.num_recall_samples, **endpoint=True**)`.

File: src/av2/evaluation/detection/eval.py [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+    dts: pd.DataFrame,
+    gts: pd.DataFrame,
+    cfg: DetectionCfg,
+) -> pd.DataFrame:
+    """Calculate and print the 3D object detection metrics.
+
+    Args:
+        dts: (N,15) Table of detections.
+        gts: (M,15) Table of ground truth annotations.
+        cfg: Detection configuration.
```
Comment (jhonykaesemodel) [inline src/av2/evaluation/detection/eval.py:173]: I'd add meaning for 15 here (e.g., where 15 is the number of attributes (e.g., x, y, z).

**Review by jhonykaesemodel — COMMENTED** [review]
Looks pretty good overall, Ben. Left some comments.

File: src/av2/evaluation/detection/utils.py (hunk 7 above)
Comment (benjaminrwilson) [reply src/av2/evaluation/detection/utils.py:71]: I really like that --- good call.

**Review by jhonykaesemodel — APPROVED** [review]
LGTM! Thanks, Ben!

**Pull request merged**